### PR TITLE
[hybrid performance] Grad fuse for gradient merge under pipeline mode

### PR DESCRIPTION
**Pull request by FeixLiu**
### PR types
Performance optimization

### PR changes
Others

### Describe
Fused gradient merge under pipeline mode

_The following test are using Ernie 3.0 model on 8 V100 GPUs, with PP=2, MP=2 and DP=2_
#### Throughput compared tokens/s (increase compared with baseline)
|Baseline speed|fp16 allreduce speed|grad fuse speed|fp16 allreduce + grad fuse speed | fp16 allreduce + optimize cast + grad fuse speed|
|:--:|:--:|:--:|:--:|:--:|
| 34285 | 35228 (+2.8%) | 36144 (+5.5%) | 35596 (+3.9%)|39145(+14.2%)|

#### Loss compared between baseline and fp16 allreduce
base_fp16_loss
base_fp16_diff

#### Loss compared between baseline and grad fuse
base_grad_fuse_loss
base_grad_fuse_diff

#### Loss compared between baseline and fp16 allreduce with grad fuse
base_fp16_grad_fuse_loss
base_fp16_grad_fuse_diff

#### Loss compared between baseline and fp16 allreduce, optimizer cast with grad fuse
Screen Shot 2021-08-19 at 6 54 11 PM
Screen Shot 2021-08-19 at 6 53 59 PM

#### NPU Loss diff （By Peng Liu)
db6d5bb116d048178f5f41f94b3c8d34

### CODE REVIEW

File: python/paddle/fluid/optimizer.py [hunk 1]
```diff
@@ -5171,6 +5178,241 @@ def _accumulate_gradients(self,
...
+                grad_param_segments[-1][1].append(real_param)
+                grad_param_segments[-1][2].append(merged_grad_var)
+                cur_size += tmp_size
+
+        fused_gradients = []
+        fused_merged_gradients = []
+        # create fused vars for grad and param
+        for grad_param_segment in grad_param_segments:
+            grad_segment = grad_param_segment[0]
+            fused_grad_segment = grad_param_segment[2]
```
Comment (wangxicoding) [inline python/paddle/fluid/optimizer.py]: 叫做merged_grad_segment吧

File: python/paddle/fluid/optimizer.py [hunk 2]
```diff
@@ -5171,6 +5178,241 @@ def _accumulate_gradients(self,
...
+        last_dtype = None
+        # split the grad based on dtype and fused size
+        for grad, param in grad_param_pairs:
+            real_grad = main_block.var(grad)
+            # create the gradient merged var for each grad
+            merged_grad_var = main_block.create_var(
+                name=param + core.grad_var_suffix() + merged_suffix,
+                dtype=dtype,
+                shape=real_grad.shape,
+                persistable=True,
```
Comment (wangxicoding) [inline python/paddle/fluid/optimizer.py:5232]: 后面可以试一下，非persistable应该也可以

File: python/paddle/fluid/optimizer.py [hunk 3]
```diff
@@ -5171,6 +5178,241 @@ def _accumulate_gradients(self,
...
+                stop_gradient=False)
+            # keep the '.cast_fp16' info in the fuse var name
+            fused_merged_grad_name_prefix = 'FusedMergedGrad.cast_fp16.' if \
+                fused_grad_segment[0].dtype == paddle.float16 else 'FusedMergedGrad'
+            fused_merged_grad_name = fused_merged_grad_name_prefix + '_{}'.format(
+                fused_grad_segment[0].name)
+            fused_merged_grad = main_block.create_var(
+                name=fused_merged_grad_name,
+                dtype=fused_grad_segment[0].dtype,
+                persistable=True,
```
Comment (wangxicoding) [inline python/paddle/fluid/optimizer.py:5271]: 非persistable后面可以试一下

File: python/paddle/fluid/optimizer.py [hunk 4]
```diff
@@ -5171,6 +5178,241 @@ def _accumulate_gradients(self,
...
+            params = grad_param_segments[i][1]
+            merged_grads = grad_param_segments[i][2]
+            main_block._insert_op_without_sync(
+                first_back_op_idx + offset,
+                type="coalesce_tensor",
+                inputs={"Input": params},
+                outputs={"Output": grads,
+                         "FusedOutput": fused_grad},
+                attrs={
+                    "user_defined_size_of_dtype": 2,
```
Comment (wangxicoding) [inline python/paddle/fluid/optimizer.py:5312]: 这个2的理由这里再给一下注释

File: python/paddle/fluid/optimizer.py [hunk 5]
```diff
@@ -5171,6 +5178,241 @@ def _accumulate_gradients(self,
...
+                    attrs={
+                        'in_dtype': fused_grad.dtype,
+                        'out_dtype': cast_grad_var.dtype,
+                        self._op_role_key: self._op_role.Backward,
+                    })
+                offset += 1
+                fused_grad = cast_grad_var
+            main_block._insert_op(
+                index=first_opt_op_idx + offset,
+                type='sum',
```
Comment (wangxicoding) [inline python/paddle/fluid/optimizer.py:5372]: 要不这个改成elementwise_add，理论上会快一丢丢

File: python/paddle/fluid/optimizer.py [hunk 6]
```diff
@@ -5171,6 +5178,241 @@ def _accumulate_gradients(self,
...
+                real_grad = main_block.var(grad)
+                fp16_grad_name = param + core.grad_var_suffix() + '@MERGED@FP16'
+                assert main_block.has_var(fp16_grad_name)
+                fp16_grad = main_block.var(fp16_grad_name)
+                fp32_grad_name = param + core.grad_var_suffix() + '@MERGED'
+                fp32_grad = main_block.create_var(
+                    name=fp32_grad_name,
+                    dtype=paddle.float32,
+                    shape=real_grad.shape,
+                    persistable=True,
```
Comment (wangxicoding) [inline python/paddle/fluid/optimizer.py]: 这个，明确不需要persistable

File: python/paddle/fluid/tests/unittests/test_fleet_sharding_meta_optimizer.py [hunk 7]
```diff
@@ -1050,6 +1050,189 @@ def test_hybrid_with_pp_dp_amp_fp16allreduce_optimize_offload(self):
...
+            'elementwise_mul_grad', 'mean_grad', 'cross_entropy_grad2',
+            'softmax_grad', 'elementwise_add_grad', 'cast', 'mul_grad',
+            'tanh_grad', 'elementwise_add_grad', 'mul_grad', 'tanh_grad',
+            'elementwise_add_grad', 'mul_grad', 'tanh_grad',
+            'elementwise_add_grad', 'mul_grad', 'c_sync_calc_stream', 'send_v2',
+            'sum', 'cast', 'sum', 'c_allreduce_sum', 'c_allreduce_sum', 'cast',
+            'cast', 'cast', 'cast', 'cast', 'cast', 'cast', 'cast',
+            'c_sync_comm_stream', 'check_finite_and_unscale', 'cast',
+            'c_allreduce_max', 'cast', 'update_loss_scaling', 'momentum',
+            'momentum', 'momentum', 'momentum', 'momentum', 'momentum',
```
Comment (wangxicoding) [inline python/paddle/fluid/tests/unittests/test_fleet_sharding_meta_optimizer.py]: 单测没有本地测试过吧。

File: python/paddle/fluid/optimizer.py (hunk 5 above)
Comment (wangxicoding) [reply python/paddle/fluid/optimizer.py:5372]: 之后改吧，否则单测改一大堆。。

**Review by wangxicoding — COMMENTED** [review]
另外再测一组 optimize_cast + fp16_allreduce + fuse_grad_merge 吧

File: python/paddle/fluid/optimizer.py (hunk 6 above)
Comment (FeixLiu) [reply python/paddle/fluid/optimizer.py]: 👌

File: python/paddle/fluid/optimizer.py (hunk 1 above)
Comment (FeixLiu) [reply python/paddle/fluid/optimizer.py]: 👍

File: python/paddle/fluid/optimizer.py (hunk 5 above)
Comment (FeixLiu) [reply python/paddle/fluid/optimizer.py:5372]: hh，好的～

File: python/paddle/fluid/tests/unittests/test_fleet_sharding_meta_optimizer.py (hunk 7 above)
Comment (FeixLiu) [reply python/paddle/fluid/tests/unittests/test_fleet_sharding_meta_optimizer.py]: optimizer cast与fused grad merge功能有冲突，will update

File: python/paddle/fluid/optimizer.py (hunk 4 above)
Comment (FeixLiu) [reply python/paddle/fluid/optimizer.py:5312]: done

**Comment by FeixLiu** [comment]
> 另外再测一组 optimize_cast + fp16_allreduce + fuse_grad_merge 吧

正在跑～～

File: python/paddle/fluid/optimizer.py (hunk 2 above)
Comment (FeixLiu) [reply python/paddle/fluid/optimizer.py:5232]: 试验中，如果可行，下一个pr改

File: python/paddle/fluid/optimizer.py (hunk 3 above)
Comment (FeixLiu) [reply python/paddle/fluid/optimizer.py:5271]: 试验中，如果可行，下一个pr改

**Comment by FeixLiu** [comment]
还有优化空间，grad和param按dtype分类之后再fuse，可以减少coalesce op个数以及fused var的个数。下一个pr可以继续优化。

**Review by wangxicoding — APPROVED** [review]
LGTM

**Pull request merged**